### PR TITLE
add testing file for saved dqn

### PR DESCRIPTION
**Pull request by zhouzypaul**
the testing file trains for 10000 steps (not exactly 100 episodes), and evaluates 100 different times, all of which give a score of 195-200

### CODE REVIEW

**Comment by camall3n** [comment]
I'm a little confused. Why is it using both the training and the testing environment? We don't need to train more; we only need the evaluations for this part.

**Comment by zhouzypaul** [comment]
my bad, that's some redundant code from `train.py` that I haven't removed. I pushed another version, and the results are the same.

File: actgen/test.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+logging.basicConfig(level=logging.INFO)
+
+
+class Trial:
+    def __init__(self, test=True):
+        args = self.parse_args()
+        self.params = self.load_hyperparams(args)
+        # hyper params to set up testing
+        self.params['max_env_steps'] = 10000
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        self.params['eval_every_n_steps'] = 100
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self.params['epsilon_final'] = 0
+        self.params['n_eval_episodes'] = 10
```
Comment (camall3n) [inline actgen/test.py]: = 100

File: actgen/test.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        self.params['replay_warmup_steps'] = 0
+        if self.params['test'] or test:
```
Comment (camall3n) [inline actgen/test.py]: The only thing I'd use test mode for here is to skip loading the q network. That way we can make sure the rest of the script runs in the CI system, but we don't need to save the models in git

File: actgen/test.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            self.params['test'] = True
+            self.params['max_env_steps'] = 1000
+            self.params['eval_every_n_steps'] = 100
+            self.params['epsilon_decay_period'] = 250
```
Comment (camall3n) [inline actgen/test.py]: = 0

And move up above the test mode section

File: actgen/test.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        args = self.parse_args()
+        self.params = self.load_hyperparams(args)
+        # hyper params to set up testing
+        self.params['max_env_steps'] = 10000
+        self.params['eval_every_n_steps'] = 100
+        self.params['epsilon_final'] = 0
+        self.params['n_eval_episodes'] = 10
+        self.params['replay_warmup_steps'] = 0
+        if self.params['test'] or test:
+            self.params['test'] = True
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            self.params['max_env_steps'] = 1000
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            self.params['eval_every_n_steps'] = 100
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            self.params['epsilon_decay_period'] = 250
+            self.params['n_eval_episodes'] = 2
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            self.params['replay_warmup_steps'] = 50
```
Comment (camall3n) [inline actgen/test.py]: remove

File: actgen/test.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            while not done:
+                a = self.agent.act(s, testing=True)
+                sp, r, done, _ = self.test_env.step(a)
+                s, G, t = sp, G + r, t + 1
+            ep_scores.append(G.detach())
+        avg_episode_score = (sum(ep_scores)/len(ep_scores)).item()
+        logging.info("Evaluation: step {}, average score {}".format(
+            step, avg_episode_score))
+
+    def run(self):
```
Comment (camall3n) [inline actgen/test.py]: remove entire run method

File: actgen/test.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        :return:
+        """
+        for step in tqdm(range(self.params['max_env_steps'])):
+            utils.every_n_times(self.params['eval_every_n_steps'], step, self.evaluate, step)
+        self.teardown()
+
+
+def main(test=False):
+    trial = Trial(test=test)
+    trial.run()
```
Comment (camall3n) [inline actgen/test.py]: replace with `trial.evaluate(0)`

File: actgen/test.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+
+class Trial:
+    def __init__(self, test=True):
+        args = self.parse_args()
+        self.params = self.load_hyperparams(args)
+        # hyper params to set up testing
+        self.params['max_env_steps'] = 10000
+        self.params['eval_every_n_steps'] = 100
+        self.params['epsilon_final'] = 0
```
Comment (camall3n) [inline actgen/test.py]: keep

File: actgen/test.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        self.params['n_eval_episodes'] = 10
+        self.params['replay_warmup_steps'] = 0
```
Comment (camall3n) [inline actgen/test.py]: keep

File: actgen/test.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+from .agents import RandomAgent, DQNAgent
+
+logging.basicConfig(level=logging.INFO)
+
+
+class Trial:
+    def __init__(self, test=True):
+        args = self.parse_args()
+        self.params = self.load_hyperparams(args)
+        # hyper params to set up testing
```
Comment (camall3n) [inline actgen/test.py]: We might want to `assert args.agent == 'dqn'`

File: actgen/test.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        test_env = wrap.FixedDurationHack(test_env)
+        test_env = wrap.DuplicateActions(test_env, self.params['duplicate'])
+        test_env = wrap.TorchInterface(test_env)
+        seeding.seed(1000+self.params['seed'], gym, test_env)
+        self.test_env = test_env
+
+        if self.params['agent'] == 'random':
+            self.agent = RandomAgent(test_env.observation_space, test_env.action_space)
+        elif self.params['agent'] == 'dqn':
+            self.agent = DQNAgent(test_env.observation_space, test_env.action_space, self.params)
```
Comment (camall3n) [inline actgen/test.py]: Let's only keep the DQNAgent

File: actgen/test.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        # load saved model
+        self.agent.q.load("results/qnet_best.pytorch")
+        self.best_score = -np.inf
```
Comment (camall3n) [inline actgen/test.py]: remove `self.best_score`

File: actgen/test.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        test_env = wrap.TorchInterface(test_env)
+        seeding.seed(1000+self.params['seed'], gym, test_env)
+        self.test_env = test_env
+
+        if self.params['agent'] == 'random':
+            self.agent = RandomAgent(test_env.observation_space, test_env.action_space)
+        elif self.params['agent'] == 'dqn':
+            self.agent = DQNAgent(test_env.observation_space, test_env.action_space, self.params)
+        # load saved model
+        self.agent.q.load("results/qnet_best.pytorch")
```
Comment (camall3n) [inline actgen/test.py]: wrap this in `if not self.params['test']`

**Review by camall3n — CHANGES_REQUESTED** [review]
Okay, this makes more sense. Have a look at my suggestions inline; I think we can streamline even further.

I expect this code will run fast enough that we can test it in the CI system, but I still think we should avoid storing model files in git. Let's use the `--test` mode to skip loading from the model file, and have the CI system only check that the program runs, not that it achieves a certain performance.

File: actgen/test.py (hunk 6 above)
Comment (zhouzypaul) [reply actgen/test.py]: is this meant for the line `self.params['test'] = True`? I thought we actually want to keep this for using `self.params['test']` on whether to load saved model or not

Comment (camall3n) [reply actgen/test.py]: Ah, you're right that one should stay.

**Comment by zhouzypaul** [comment]
thanks so much for all the detailed suggestions! I've made the changes.
Oops looks like it's failing again, but I'll fix it.

**Comment by zhouzypaul** [comment]
I'm not sure what's failing the tests currently, I think it might be pytest is doing something weird. 
I can run `python3 -m actgen.test --test` without the saved model, as well as `actgen.test.main(test=True)`. 

But the pytest just fails. (pytest also keeps trying to generate the saved model file when it's not found, which is also kind of weird)

**Review by camall3n — APPROVED** [review]

**Pull request merged**